pkg/syncer/controller/controller.go

```diff
@@ -0,0 +1,186 @@
+	"github.com/kcp-dev/edge-mc/pkg/syncer/syncers"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"k8s.io/klog/v2"
+)
+
+const (
+	controllerName = "kcp-edge-syncconfig-controller"
+)
+
+// NewSyncConfigController returns a controller watching a [edge.kcp.io.EdgeSyncConfig] and update down/up syncer,
+func NewSyncConfigController(
+	logger klog.Logger,
+	syncConfigClient edgev1alpha1typed.EdgeSyncConfigInterface,
+	syncConfigInformer edgev1alpha1informers.EdgeSyncConfigInformer,
+	syncConfigName string,
+	upSyncer syncers.SyncerInterface,
+	downSyncer syncers.SyncerInterface,
+) (*controller, error) {
+	rateLimitter := workqueue.NewItemFastSlowRateLimiter(5*time.Second, 30*time.Second, 1000)
+	c := &controller{
+		queue:                       workqueue.NewNamedRateLimitingQueue(rateLimitter, controllerName),
+		syncConfigLister:            syncConfigInformer.Lister(),
+		syncConfigInformerHasSynced: syncConfigInformer.Informer().HasSynced,
+		syncConfigClient:            syncConfigClient,
+		upSyncer:                    upSyncer,
+		downSyncer:                  downSyncer,
+	}
+	syncConfigInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err != nil {
+				return false
+			}
+			_, name, err := cache.SplitMetaNamespaceKey(key)
+			if err != nil {
+				return false
+			}
+			return name == syncConfigName
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) { c.enqueueSyncConfig(obj, logger) },
+			UpdateFunc: func(old, obj interface{}) { c.enqueueSyncConfig(obj, logger) },
+			DeleteFunc: func(obj interface{}) { c.enqueueSyncConfig(obj, logger) },
+		},
+	})
+
+	return c, nil
+}
+
+// controller is a control loop that watches EdgeSyncConfig.
+type controller struct {
+	queue                       workqueue.RateLimitingInterface
+	syncConfigLister            edgev1alpha1listers.EdgeSyncConfigLister
+	syncConfigInformerHasSynced cache.InformerSynced
+	syncConfigClient            edgev1alpha1typed.EdgeSyncConfigInterface
+	upSyncer                    syncers.SyncerInterface
+	downSyncer                  syncers.SyncerInterface
+}
+
+// Ready returns true if the controller is ready to return the GVRs to sync.
+// It implements [informer.GVRSource.Ready].
+func (c *controller) Ready() bool {
+	return c.syncConfigInformerHasSynced()
+}
+
+func (c *controller) enqueueSyncConfig(obj interface{}, logger klog.Logger) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	shared.WithQueueKey(logger, key).V(2).Info("queueing SyncConfig")
+	c.queue.Add(key)
+}
+
+// Start starts the controller workers.
+func (c *controller) Start(ctx context.Context, numThreads int) {
```

Comment:
A better name for this method is "Run".  The word "start" does not imply waiting until completion, but "run" does.